streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -596,6 +596,68 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
       ZStream.fromQueue(queue).unTake
     }
 
+  /**
+   * Creates a stream that passes its elements through a queue that drops elements
+   * (dropping or sliding). It implements a promise that signals abrupt completion
+   * of the stream due to an error or the end of the stream. After the promise has
+   * been completed, it continues to offer the current (at the time) contents of the
+   * queue, until it has been exhausted, at which point it finishes with the signal
+   * from the promise.
+   */
+  private final def bufferSignal[E1 >: E, A1 >: A](queue: Queue[Take[E1, A1]]): ZStream[R, E1, A1] =
+    ZStream.managed {
+      for {
+        signal <- Promise.make[Nothing, Take[E, Nothing]].toManaged_
+        fiber  <- signal.await.toManaged_.fork
+        _ <- self
+              .foreachManaged(a => queue.offer(Take.Value(a)).unit)
+              .foldCauseM(
+                c => signal.succeed(Take.Fail(c)).unit.toManaged_,
+                _ => signal.succeed(Take.End).unit.toManaged_
+              )
+              .fork
+      } yield (queue, fiber)
+    }.flatMap {
+      case (queue, fiber) =>
+        val pull = queue.poll.flatMap {
+          _.fold[Pull[Any, Nothing, Take[E1, A1]]](
+            queue.take.raceWith(fiber.join)(
+              (l, _) => l.fold(Pull.halt, Pull.emit),
```

Comment:
This leaks the joining fiber every time, so make sure you cancel it

Reply:
Isn't this the same fiber passed from the previous `flatMap`? It will get collected when the `Managed` scope ends. Or am I wrong?

Reply:
Or `raceWith` constructs another fiber always?